### ISSUE TITLE
DST Staging Review (Consider): Voice interaction breaks down

### ISSUE THREAD

**Issue by shiragoodman** [department-of-veterans-affairs/vets-design-system-documentation#3085]
_Need help? Please review [how to read a Staging Review ticket](https://depo-platform-documentation.scrollhelp.site/collaboration-cycle/anatomy-of-a-staging-review-issue-ticket). Tag `@platform-governance-team-members` on Slack if you need further assistance._

## Design System Staging Information
**Component:** File Input Component
**Staging Review ticket:** [Staging Review: (DS) Converted File Input Component](https://github.com/department-of-veterans-affairs/va.gov-team/issues/77552)

## Findings details
The details of this ticket are for consideration only. No Staging Review issue was found.
**High-priority:** No
**Collab Cycle Reviewer:** @briandeconinck (Accessibility)

### Description

I know these design choices are coming from the USWDS and this was raised with them at some point last year, but I'll flag it again in hopes of some additional conversation.

There's some styling here that's great and intuitive for sighted mouse/keyboard users, but the interaction breaks down for sighted voice command software users. Moreover, it violates principles of material honesty --- which is maybe defensible, but I don't think is actually necessary.

Visually, the component looks like a `label`, a drag/drop target for mouse users, and a "choose from folder" link inside that target that can be clicked to select a file from your computer's file manager app.

![Component with label "Select a file to upload," allowed file format hint text, and a dotted line rectangle with text "Drag files here or choose from folder" inside, with "choose from folder" blue and underlined.](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/working-with-vsp/vsp-collaboration-cycle/staging-review-images/Brian-77552-2.png?raw=true)

When you tab into the component, you get the appearance of the drag/drop target receiving focus.

![Component with label "Select a file to upload," hint text, and a dotted line "Drag files here" target, with a yellow focus indicator around the "Drag files here" target.](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/working-with-vsp/vsp-collaboration-cycle/staging-review-images/Brian-77552-3.png?raw=true)

But the "choose from folder" link isn't actually a link --- it's just a non-interactive `span`. And in fact, the drag/drop target doesn't actually receive focus, or even play any role at all in the file upload. Instead, the drag/drop target is a non-interactive `div` that's covering up the real `input` so as to conflate them visually.
```
<div class="file-input-target">
	<div class="file-input-box"></div>
	<div class="file-input-instructions">
		<span class="file-input-drag-text">Drag files here or </span>
		<span class="file-input-choose-text">choose from folder</span>
	</div>
</div>
```

Below, I've tweaked the CSS to make the real `input` visible again, with browser-default styling.

![Component with label "Select a file to upload," allowed file format hint text, a browser-default file upload element consisting of a "Choose file" button and "No file chosen" text wrapped in a red outline, and a dotted line rectangle with text "Drag files here or choose from folder" inside, with "choose from folder" blue and underlined.](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/working-with-vsp/vsp-collaboration-cycle/staging-review-images/Brian-77552-4.png?raw=true)

This is almost totally fine!
- For mouse users, you can drag a file into the target (and you'll also happen to be above the real `input`), or you can click "choose from folder" (and you'll also happen to click the real `input`).
- For keyboard users, you can tab to where you'd expect to tab to and press enter (because even though it looks like you've tabbed to the drag/drop target, the real `input` is actually in focus).

The problem is for voice users. There's nothing a voice user can say that will successfully target the `input`.
- "Click choose from folder" doesn't work, because "choose from folder" isn't an interactive element.
- "Click Drag files here or choose from folder" also doesn't work, for the same reasons.
- And, because nothing is easy, even "Click Select a file to upload" doesn't work --- even though it seems like it should work the way other inputs work. As best I can tell, here's what's happening:
   - Voice command software looks for matches in the accessibility tree
   - The browser labels `input type="file"` as a `button` in the accessibility tree
   - "Select a file to upload" is the label associated with that `button`
   - The browser renders the input as a button, and in rendering it adds its own label (eg. "Choose file" in Chrome) that overrides that "Select a file to upload" label, so now the input is labeled with whatever the browser picked (eg. "Choose file" in Chrome)
   - The magic voice command should be "Click Choose file" or whatever it is in your browser, which a voice user wouldn't know to try because any text that says "Choose file" is covered up by the drag/drop target.
      - Except "Click Choose file" doesn't seem to work when I test either, so... ???

Testing with Dragon, the only ways I'm able to successfully upload a file are either via direct keystroke commands ("Press tab") or mouse commands (eg. MouseGrid).

And truthfully, most voice users are probably used to doing that on lots of websites --- and possibly on all file uploads, which may just be broken for voice across all browsers (?!?!). The actual impact to voice users may be minimal.

But I think there may be a way to make this work.

## Recommended action
Martin Mengele's [A styled accessible file upload](https://accessabilly.com/a-styled-accessible-file-upload/) is structured very differently than what USWDS is doing, but I think there may be a way to synthesize this approach with the USWDS look/feel.

Mengele uses two inputs --- a hidden `type="file"` and a visible `type="text", and uses a button (or rather a `span role="button"` inside a `label`) to control them together. It's kind of bonkers? But I've confirmed that it works with Dragon.

I think the "choose from folder" pseudo-link may be a good candidate for the `button` here. (Ideally it would be styled as a button, but honestly I'd let the material dishonesty slide if we can solve the rest!) That way, the voice command "Click choose from folder" would target an element that's actually interactive.

The way Mengele uses `label` isn't compatible with what's in the USWDS component, but I think a `fieldset`/`legend` could come to our rescue here. Wrap the whole thing in a `fieldset` and make "Select a file to upload" and the hint text the `legend` and that frees up the `label` element to be used as needed elsewhere.

I'm not 100% confident that this will all work, but I'd like to recommend it as something to explore further (and I'm also very happy to help with the coding and testing as much as I'm able).

## References

- **Accessibility Defect Severity:** [4: Low. Consider fixing or exploring in 2-4 sprints post-launch.](https://depo-platform-documentation.scrollhelp.site/developer-docs/accessibility-defect-severity-rubric#Accessibilitydefectseverityrubric-a11y-defect-4)
- **WCAG Success Criteria:** N/A
- **Modality:** Voice
- **Design System Pattern or Template:** N/A
- **Design System Foundation:** N/A
- **Content Style Guide:** N/A
- **Context:** N/A

---

### Next Steps for DST

- Close the ticket when the issue has been resolved or validated by your Product Owner
- If your team has additional questions or needs Governance help validating the issue, please comment on the ticket
- If this ticket has a `high-priority` label, please address as soon as possible so VFS teams are not impacted
- If this ticket has a `consider` label, please consider for future implementation

**Comment by caw310** [comment]
Related USWDS issue; https://github.com/uswds/uswds/issues/5616

**Comment by caw310** [comment]
This will require review and reworking design with Jeana, Dan and Barb before development can work on this.

**Comment by caw310** [comment]
@amyleadem @jeana-adhoc recommend closing as it is accessible.
